### ISSUE TITLE
Cannot use references with install()

### ISSUE THREAD

**Issue by krojew** [google/fruit#66]
Given the situation:

`fruit::Component<...> getFoo(const Widget &) { ... }`

it's impossible to pass a std::reference_wrapper to the install method:

```
Widget w;
fruit::createComponent().install(getFoo, std::cref(w));
```

It would be great if we could use references in such situations.

**Comment by poletti-marco** [comment]
Hi, could you please explain what's your use case? E.g. in your example, why don't you just pass `widget`?

**Comment by krojew** [comment]
This is impossible if Widget is non-copyable.

**Comment by poletti-marco** [comment]
Yes, that's true. If possible, in this case I would recommending injecting the widget instead.

* If the widget is an object that outlives the injector (e.g. a constant), you could bind it with `bindInstance`.
* Otherwise, as you can imagine, creating a local, non-copyable `Widget` local var as in your example and passing a reference means that the referenced `Widget` object would be destroyed at the end of that method, so you'd hit undefined behavior.

If there are multiple (but finite) instances of `Widget` in a single injector, you can bind them using [annotated injection](https://github.com/google/fruit/wiki/quick-reference#annotated-injection), perhaps using a templated `get*Component` function parameterized on the tag type.

For example:

    template <typename Tag>
    fruit::Component<
        fruit::Required<fruit::Annotated<Tag, Widget>>, 
        fruit:Annotated<Tag, Foo>
    > getFooComponent() {
        return fruit::createComponent()
            .registerProvider<fruit:Annotated<Tag, Foo>(fruit::Annotated<Tag, Widget>&)>([](Widget& w) {
                return Foo(w);
            });
    }
    
    struct MainWidget {};

    struct MainWidgetImpl : public Widget {
        INJECT(MainWidgetImpl()) = default;
        ...
    };
    
    fruit::Component<fruit::Annotated<MainWidget, Foo>> getMainWidgetFooComponent() {
        return fruit::createComponent()
            .install(getFooComponent)
            .bind<fruit::Annotated<MainWidget, Widget>, MainWidgetImpl>();
    }

Not sure if that solves your problem though, because you're asking about a solution rather than a problem (see [XY problem](https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem)).

**Comment by krojew** [comment]
Let me get into more detail on the problem I have. I hope it will make things clearer.

My injections are dynamic - there's a configuration class which determines what should be the actual implementation to inject. This configuration class is non-copyable and should not be. With the current design of fruit, I need to pass it to the get*Component() functions in order for them to know what to actually inject (in practice it's more complex, but this simplification should be descriptive enough). At that point I know the config object is alive and will outlive the injector. Yet, there's no way to pass a reference to it; I need to resort to pointers, which by nature allow the nonexistence of the object. This implies my get*Component functions have a precondition (config is alive), but also allow that precondition to be unintentionally violated (config parameter is a nullable pointer). This happens only because of the inability of fruit ot pass a reference.

**Comment by poletti-marco** [comment]
Thanks for the info.
One more question: is there a unique configuration object in a given process (or at least in a given thread), or can you have multiple of these per process/thread?

If there's a single one per process, you could have it as a static object (possibly initialized later) and access that in the `get*Component` functions (probably via a getter function).

If there's a single one per thread, you could do something similar with a thread-local "static" object.

Otherwise, you could define a copyable "pointer wrapper" type that satisfies the requirements on parameters [here](https://github.com/google/fruit/wiki/quick-reference#templated-and-parametrized-components) like:

struct ConfigRef {
    Config* config;
    ConfigRef(Config& config) : config(&config) {}
    bool operator==(const ConfigRef& other) const {
        return config == other.config;
    }
};

namespace std {
    template <> 
    struct hash<ConfigRef> {
        size_t operator()(const ConfigRef& x) const {
            return hash<Config*>()(x.config);
        }
    };
}

Since the only constructor takes a reference, you're guaranteed that ConfigRef.config is never null (so you don't need to document/check this in `get*Component` functions), while at the same time it's usable as a parameter of `get*Component` functions.

WDYT?

**Comment by krojew** [comment]
There can be multiple config objects. Creating the wrapper doesn't really solve the logical problem - you still have a nullable pointer, while the precondition is for the object to exists. The only difference is a wrapper around it. This might not seem like a big deal, but it has the consequence of requiring changes to the interface which stem only from technical problems with a 3rd party dependency.

**Comment by poletti-marco** [comment]
> Creating the wrapper doesn't really solve the logical problem - you still have a nullable pointer, while the precondition is for the object to exists.

I'm not sure I understand.
If you have a function like:

    fruit::Component<...> getFooComponent(ConfigRef configRef) {
        ...
    }

Then you don't need to document nor check the nullness of the pointer in this function. When a caller tries to call this, they'll either pass an existing ConfigRef object or need to construct one with a Config&.

(let's assume that the field in the struct is actually private, so there's no way for other code to set a null value)

>This might not seem like a big deal, but it has the consequence of requiring changes to the interface which stem only from technical problems with a 3rd party dependency.

Only the interface of the `get*Component` function needs to be changed. If your library has a public interface that doesn't mention Fruit, then you can take a Config& there if you want, and convert to a ConfigRef when you call your `get*Component` functions in order to create the injector.
So in this case it's not requiring changes to the interface.

If you do decide to expose the `get*Component` functions (having Fruit as part of your public interface) then Fruit is not just a "3rd party dependency", and you'd already need to satisfy the requirements of the Fruit API (e.g. that parameter types of installed `get*Component` methods need to be assignable, which references aren't).
So in this case it's also not requiring changes to the interface (while it *is* requiring a different interface), since this was never supported by the Fruit API.

The rules on what parameter types are allowed in `get*Component` functions are already non-trivial, I'm wary of complicating them further by special-casing `std::reference_wrapper` there unless there's a strong need (while AFAICT here there's an easy alternative that you can use in this case).

**Comment by krojew** [comment]
By not solving the problem I meant the object which is not supposed to be nullable (Config in this example), can still be null when using the wrapper. Only the wrapper is not nullable. In this regard nothing has changed. The precondition requires the config to not be null, but fruit doesn't allow such preconditions in get() functions.

**Comment by poletti-marco** [comment]
> By not solving the problem I meant the object which is not supposed to be nullable (Config in this example), can still be null when using the wrapper.

It can't if the field is private, AFAICT. Maybe it's because I put it as public in the code above (it was for simplicity). But if you make it private:

```
class ConfigRef {
private:
    Config* config;
public:
    ConfigRef(Config& config) : config(&config) {}
    Config& getConfig() {
        return *config;
    }
    const Config& getConfig() const {
        return *config;
    }
    bool operator==(const ConfigRef& other) const {
        return config == other.config;
    }
};

namespace std {
    template <> 
    struct hash<ConfigRef> {
        size_t operator()(const ConfigRef& x) const {
            return hash<Config*>()(x.getConfig());
        }
    };
}
```

And you use `ConfigRef` as param to a `get*Component` function, then there's no way that a null Config can be passed to it, and there's no precondition to document, right?

**Comment by krojew** [comment]
That's one workaround. Yet, I would be great if no workaround was needed.
